crates/bevy_reflect/src/from_reflect.rs

```diff
@@ -12,7 +13,29 @@ use crate::{FromType, Reflect};
 /// reflected value.
 pub trait FromReflect: Reflect + Sized {
     /// Constructs a concrete instance of `Self` from a reflected value.
-    fn from_reflect(reflect: &dyn Reflect) -> Option<Self>;
+    fn from_reflect(reflect: &dyn Reflect) -> Result<Self, FromReflectError>;
+}
+
+/// An enum representing all the kinds of types that a value implementing [`Reflect`] could be.
+/// This enum is primarily used in [`FromReflectError`].
+#[derive(Debug)]
+pub enum ReflectType {
+    Struct,
+    TupleStruct,
+    Tuple,
+    List,
+    Array,
+    Map,
+    Enum,
+    Value,
+}
+
+/// An Error for failed conversion of reflected type to original type in [`FromReflect::from_reflect`].
+#[derive(Error, Debug)]
+#[error("The reflected type of `{}` cannot be converted to {:?}", .from_type_name, .to_type)]
+pub struct FromReflectError<'a> {
```

Comment:
Imo this should be an enum since we have multiple reasons this method could fail:
1. If the `dyn Reflect` is not the correct `ReflectType`
2. If the `dyn Reflect` is missing a required field
3. If a sub-field within the `dyn Reflect` fails their own `FromReflect::from_reflect`

And it would probably be more beneficial to the user to let them know what caused the actual failure.

The third one might be a little difficult to track haha, but we could maybe have a recursive variant for it. So in my mind we should be able to do something like:
```rust
enum FromReflectError<'a> {
    InvalidType {
        from_type_name: &'a str,
        to_type: ReflectType,
    },
    MissingField {
        from_type_name: &'a str,
        field_name: &'a str
    },
    FieldError {
        from_type_name: &'a str,
        field_name: &'a str,
        error: Box<FromReflectError<'a>>
    }
}
```

Reply:
How does this API look like? I have taken all of your considerations into account. I also decided to use respective types instead of `&str` which should be more idiomatic in Rust. For now, I have decided to call it `ReflectCategory` as it has the least baggage associated with it and it can be changed later if we want.
```rust
/// An enum representing all the categories of types that a value implementing [`Reflect`] could be.
#[derive(Debug)]
pub enum ReflectCategory {
    Struct,
    TupleStruct,
    Tuple,
    List,
    Array,
    Map,
    Enum,
    Value,
    Dynamic,
}

/// An Error for failed conversion of reflected type to original type in [`FromReflect::from_reflect`].
#[derive(Error, Debug)]
pub enum FromReflectError<'a> {
    #[error("The reflected type `{}` of category {:?} cannot be converted to type `{}` of category {:?} due to mismatched categories", 
            .from_type.type_name(), .from_type.category(), .to_type.type_name(), .to_type.category())]
    InvalidCategory {
        from_type: &'a TypeInfo,
        to_type: &'a TypeInfo,
    },
    #[error("The reflected type `{}` of category {:?} cannot be converted to type `{}` due to a missing field `{}` of type `{}`", 
            .from_type.type_name(), .from_type.category(), .to_type.type_name(), .field.name(), .field.type_name())]
    MissingNamedField {
        from_type: &'a TypeInfo,
        to_type: &'a TypeInfo,
        field: &'a NamedField,
    },
    #[error("The reflected type `{}` of category {:?} cannot be converted to type `{}` due to a missing field at index `{}` of type `{}`", 
            .from_type.type_name(), .from_type.category(), .to_type.type_name(), .field.index(), .field.type_name())]
    MissingUnnamedField {
        from_type: &'a TypeInfo,
        to_type: &'a TypeInfo,
        field: &'a UnnamedField,
    },
    #[error("The reflected type `{}` of category {:?} cannot be converted to type `{}` due to the following error in the field `{}` of type `{}`:\n{}", 
            .from_type.type_name(), .from_type.category(), .to_type.type_name(), .field.name(), .field.type_name(), .error)]
    NamedFieldError {
        from_type: &'a TypeInfo,
        to_type: &'a TypeInfo,
        field: &'a NamedField,
        error: Box<FromReflectError<'a>>,
    },
    #[error("The reflected type `{}` of category {:?} cannot be converted to type `{}` due to the following error in the field at index `{}` of type `{}`:\n{}", 
            .from_type.type_name(), .from_type.category(), .to_type.type_name(), .field.index(), .field.type_name(), .error)]
    UnnamedFieldError {
        from_type: &'a TypeInfo,
        to_type: &'a TypeInfo,
        field: &'a UnnamedField,
        error: Box<FromReflectError<'a>>,
    },
}
```

The above code assumes a function `.category()` exists on `TypeInfo` returning the associated `ReflectCategory`. I will be adding it in this PR itself.
Question: The error message in NamedFieldError and UnnamedFieldError is recursive and it will print multiple lines depending on how deep the error initiated. Is it better like this or should only first line be printed?

Reply:
Yeah I think this would work. The only consideration would be performance, since it's somewhat costly to access `TypeInfo` using `Reflect::get_type_info` or `Typed::type_info`.

Reply:
I did not know about the performance implications of `TypeInfo`. So what should be the final API? (`&TypeInfo` or `&str`).

I guess you missed this question (or maybe not, it just was not clear from previous comment):
> Question: The error message in NamedFieldError and UnnamedFieldError is recursive and it will print multiple lines depending on how deep the error initiated. Is it better like this or should only first line be printed?

Reply:
> I did not know about the performance implications of `TypeInfo`. So what should be the final API? (`&TypeInfo` or `&str`).

I'm debating this myself on #6971 haha. I think, ultimately, `TypeInfo` will just be so much more useful. I think it should be okay here considering that this is the error case, so it shouldn't be in the hot path necessarily.

I say go for it, but you could also check in with [#reflection-dev](https://discord.com/channels/691052431525675048/1002362493634629796) to see if anyone else has any other opinions.

> I guess you missed this question (or maybe not, it just was not clear from previous comment):
> 
> > Question: The error message in NamedFieldError and UnnamedFieldError is recursive and it will print multiple lines depending on how deep the error initiated. Is it better like this or should only first line be printed?

Yep, sorry, I completely missed that question! Yeah I don't think the error needs to print recursively. Just stating the field is good enough imo. If users want more info, they can manually recurse into the enum themselves.